sqlmesh/core/plan/definition.py

```diff
@@ -238,13 +227,48 @@ def missing_intervals(self) -> t.List[SnapshotIntervals]:
 
     @property
     def snapshots(self) -> t.List[Snapshot]:
-        """Gets all the snapshots in the plan/environment."""
-        return list(self.context_diff.snapshots.values())
+        """Gets all the valid snapshots in the plan/environment."""
+        return [
+            snapshot
+            for snapshot in [
+                self._get_snapshot(snapshot_name)
+                for snapshot_name in self._dag
+                if snapshot_name not in self.invalid_snapshot_names
+            ]
+            if snapshot
+        ]
 
     @property
     def new_snapshots(self) -> t.List[Snapshot]:
         """Gets only new snapshots in the plan/environment."""
-        return list(self.context_diff.new_snapshots.values())
+        return [
+            snapshot
+            for snapshot in self.snapshots
+            if snapshot.snapshot_id in self.context_diff.new_snapshots
+        ]
+
+    @property
+    def has_changes(self) -> bool:
+        modified_names = (
+            set().union(
+                *[
+                    self.context_diff.added,
+                    self.context_diff.removed,
+                    self.context_diff.modified_snapshots,
+                ]
+            )
+            - self.invalid_snapshot_names
+        )
+        return (
+            self.context_diff.is_new_environment
+            or self.context_diff.is_unfinalized_environment
+            or bool(modified_names)
+        )
+
+    def is_valid_snapshot(self, snapshot_name: t.Union[str, Snapshot]) -> bool:
+        """Checks if a snapshot is valid. Valid meaning that it is not being ignored in the plan"""
+        snapshot_name = snapshot_name if isinstance(snapshot_name, str) else snapshot_name.name
+        return snapshot_name in self._dag
```

Comment:
The docstring claims:

> Valid meaning that it is not being ignored in the plan

But I don't see invalid snapshots being excluded from `self._dag` anywhere.

Reply:
Ah, I take this back. It happens below.